### PR TITLE
Download reports UI - dynamic year dropdown added

### LINKED ISSUE #1104: Create a UI to download a spreadsheet
This is continuation of issue #884 

Create a UI to download a spreadsheet of data that CELTS requested.

### PR DESCRIPTION
**Pull request by seedyjahateh**
We created a reports UI under the Admin tab. UI includes a download button and an input bar for the academic year of the report you want to download.
The year options in the dropdown are taken from the database.

Fixes #1104

### CODE REVIEW

File: app/controllers/admin/routes.py [hunk 1]
```diff
@@ -38,6 +38,25 @@
 from app.logic.serviceLearningCourses import parseUploadedFile, saveCourseParticipantsToDatabase, unapprovedCourses, approvedCourses, getImportedCourses, getInstructorCourses, editImportedCourses
 
 from app.controllers.admin import admin_bp
+from flask import send_file, render_template, request
+from app.controllers.admin import admin_bp
```
Comment (BrianRamsay) [inline app/controllers/admin/routes.py]: Remove redundant imports

File: app/controllers/admin/routes.py [hunk 2]
```diff
@@ -38,6 +38,25 @@
...
+from app.logic.spreadsheet import createSpreadsheet
+
+
+@admin_bp.route('/admin/reports')
+def reports():
+    academic_years = Term.select(Term.academicYear).distinct().order_by(Term.academicYear.desc())
+    academic_years = list(map(lambda t: t.academicYear, academic_years))
+    return render_template("/admin/reports.html", academic_years=academic_years)
```
Comment (BrianRamsay) [inline app/controllers/admin/routes.py]: Our variables are camelCase in our application

File: app/controllers/admin/routes.py [hunk 3]
```diff
@@ -38,6 +38,25 @@
...
+
+@admin_bp.route('/download')
+def download_file():
```
Comment (BrianRamsay) [inline app/controllers/admin/routes.py]: Same with function names

File: app/controllers/admin/routes.py [hunk 4]
```diff
@@ -38,6 +38,25 @@
...
+from app.logic.spreadsheet import createSpreadsheet
+
+
+@admin_bp.route('/admin/reports')
+def reports():
+    academic_years = Term.select(Term.academicYear).distinct().order_by(Term.academicYear.desc())
+    academic_years = list(map(lambda t: t.academicYear, academic_years))
+    return render_template("/admin/reports.html", academic_years=academic_years)
+
+@admin_bp.route('/download')
```
Comment (BrianRamsay) [inline app/controllers/admin/routes.py]: Route should look related to the page that got us here (`/admin/reports`)

File: app/controllers/admin/routes.py [hunk 5]
```diff
@@ -38,6 +38,25 @@
...
+def download_file():
+    academic_year = request.args.get('academic_year', '2023-2024')  # Default academic year if not provided
+    filepath = createSpreadsheet(academic_year)
+    filepath = os.path.abspath(filepath)
```
Comment (BrianRamsay) [inline app/controllers/admin/routes.py]: Let's combine this into one line since it fits

File: app/templates/admin/reports.html [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+
+
+{% block styles %}
+  {{ super() }}
+  <style>
+      .download-container {
+          text-align: center;
+          margin-top: 50px;
+      }
+  </style>
```
Comment (BrianRamsay) [inline app/templates/admin/reports.html]: This needs to be in a css file.

File: app/templates/admin/reports.html [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+        <div class="form-group">
+            <label for="academicYear">Academic Year:</label>            
+            <select id="academicYear" name="academic_year" class="form-control" required>
+                <option value="" disabled selected>Select an academic year</option>
+                {% for year in academic_years %}
+                    <option value="{{ year }}">{{ year }}</option>
+                {% endfor %}
+            </select>
+        </div>
+        <button type="submit" class="btn btn-primary" onclick="confirmDownload()">Download Spreadsheet</button>
```
Comment (BrianRamsay) [inline app/templates/admin/reports.html]: Let's say "Download Annual Report"

Comment (BrianRamsay) [reply app/templates/admin/reports.html]: Maybe with an excel icon in the button

**Review by BrianRamsay — CHANGES_REQUESTED** [review]

File: app/controllers/admin/routes.py (hunk 1 above)
Comment (seedyjahateh) [reply app/controllers/admin/routes.py]: removed

File: app/controllers/admin/routes.py (hunk 2 above)
Comment (seedyjahateh) [reply app/controllers/admin/routes.py]: Changed names to camelCase.

File: app/controllers/admin/routes.py (hunk 3 above)
Comment (seedyjahateh) [reply app/controllers/admin/routes.py]: Done

File: app/controllers/admin/routes.py (hunk 5 above)
Comment (seedyjahateh) [reply app/controllers/admin/routes.py]: Done.

File: app/controllers/admin/routes.py (hunk 4 above)
Comment (seedyjahateh) [reply app/controllers/admin/routes.py]: How about '/admin/reports/download'?

File: app/templates/admin/reports.html (hunk 6 above)
Comment (seedyjahateh) [reply app/templates/admin/reports.html]: Used built-in bootstrap

File: app/templates/admin/reports.html (hunk 7 above)
Comment (seedyjahateh) [reply app/templates/admin/reports.html]: Changed

**Review by BrianRamsay — CHANGES_REQUESTED** [review]
The select box doesn't need to be full screen width. Let's use our form-control styles for the label and select box. 

Download button can just say Download, with the icon in front of it.

**Comment by seedyjahateh** [comment]
> The select box doesn't need to be full screen width. Let's use our form-control styles for the label and select box.
> 
> Download button can just say Download, with the icon in front of it.

Reduced the size of 'select box', used built-in bootstrap for styling the 'select box' and 'label', and renamed the download button to 'Download' and moved the Excel icon infront of the name.

**Comment by MImran2002** [comment]
suggestion: 
image
I would recommend as this is a dropdown UI to have dropdown design. The code is perfect it's just I feel a design perspective 
link: https://getbootstrap.com/docs/4.0/components/dropdowns/

kudos that it all works very well. the download button works too!

**Review by BrianRamsay — CHANGES_REQUESTED** [review]
Looks good! However, the academic year is not being used properly when the spreadsheet is generated

**Comment by WackyWeaver** [comment]
> Looks good! However, the academic year is not being used properly when the spreadsheet is generated

We found that this needs to be a whole other issue with spreadsheets.py

**Pull request merged**